monkey/agent_plugins/payloads/cryptojacker/src/memory_utilizer.py

```diff
@@ -1,14 +1,132 @@
+MEMORY_CONSUMPTION_NOP_THRESHOLD = 0.02
+# We don't want to ever use more then 90% of available memory, otherwise we risk impacting the
+# victim machines performance
+MEMORY_CONSUMPTION_SAFETY_LIMIT = 0.9
+
+logger = logging.getLogger(__name__)
 
 
 class MemoryUtilizer:
-    def __init__(self, memory_utilization: PercentLimited):
-        self._memory_utilization = memory_utilization
+    def __init__(
+        self,
+        agent_id: AgentID,
+        agent_event_publisher: IAgentEventPublisher,
+        target_utilization: PercentLimited,
+    ):
+        self._agent_id = agent_id
+        self._agent_event_publisher = agent_event_publisher
+        self._target_utilization = target_utilization
+        self._consumed_bytes = b""
+
+        self._periodic_caller = PeriodicCaller(
+            self.adjust_memory_utilization,
+            MEMORY_CONSUMPTION_CHECK_INTERVAL,
+            name="Cryptojacker.MemoryUtilizer",
+        )
+
+    @property
+    def consumed_bytes_size(self) -> int:
+        try:
+            return len(self._consumed_bytes)
+        except AttributeError:
+            # self._consumed_bytes was deleted and is currently being reinitialized return 0 while
+            # we wait.
+            return 0
 
     def start(self):
-        pass
+        logger.debug("Starting MemoryUtilizer")
+        self._periodic_caller.start()
+
+    def adjust_memory_utilization(self):
+        try:
+            memory_to_consume = self._calculate_memory_to_consume()
+            self.consume_bytes(len(self._consumed_bytes) + memory_to_consume)
+        except RuntimeError as err:
+            logger.error("Failed to adjust memory utilization: %s", err)
+
+    def _calculate_memory_to_consume(self) -> int:
+        total_virtual_memory = psutil.virtual_memory().total
+        available_virtual_memory = psutil.virtual_memory().available
+        used_virtual_memory = psutil.Process().memory_info().vms
+
+        if used_virtual_memory > total_virtual_memory:
+            raise RuntimeError("Impossible system state: Used memory is greater than total memory")
+
+        ideal_memory_to_consume = int(
+            total_virtual_memory * self._target_utilization.as_decimal_fraction()
+            - used_virtual_memory
+        )
+        maximum_memory_to_consume = int(
+            (available_virtual_memory + used_virtual_memory) * MEMORY_CONSUMPTION_SAFETY_LIMIT
+            - used_virtual_memory
+        )
+
+        # We never want to consume 100% of available memory, otherwise the OS could kill this
+        # process or one of the user's mission-critical processes. This logic limits the amount of
+        # memory we consume to 90% of available memory.
+        return min(ideal_memory_to_consume, maximum_memory_to_consume)
+
+    def consume_bytes(self, bytes_: int):
+        logger.debug(
+            f"Currently consumed: {self.consumed_bytes_size} bytes - Target: {bytes_} bytes"
+        )
+        timestamp = time.time()
```

Comment:
Why get the timestamp all the way up here? Shouldn't we do it nearer to line 92 and line 100?